riscv-page-tables/src/page_table.rs

```diff
@@ -737,6 +779,26 @@ impl<T: PagingMode> PageTableInner<T> {
             _ => Err(Error::PageNotInvalidated),
         }
     }
+
+    /// Promotes the table entry at `vaddr` as a valid leaf.
+    fn promote_leaf(
+        &mut self,
+        vaddr: RawAddr<T::MappedAddressSpace>,
+        page_size: PageSize,
+    ) -> Result<(LeafPte<T>, Page<InternalClean>)> {
+        let mut entry = PageTable::from_root(self).entry_for_addr_mut(vaddr);
+        use TableEntryType::*;
+        while let Table(t) = entry {
+            if t.level().leaf_page_size() != page_size {
+                entry = t.table().entry_for_addr_mut(vaddr);
+            } else {
+                let table_addr = t.table_addr();
+                let page: Page<InternalDirty> = unsafe { Page::new(table_addr) };
+                return Ok((t.leaf(), page.clean()));
+            }
+        }
+        Err(Error::UpdatePteAsLeaf)
```

Comment:
`PteNotPromotable` might be a more descriptive error code

Reply:
:+1: